fix: reconnect terminal on non-modified key presses

code-asher opened the pull request:
Fixes #8547 
Fixes #6732

code-asher committed:
Listen to web terminal keydown on capture

Instead of bubbling.  I think maybe what happens here is that xterm is
capturing key presses and preventing the event from bubbling?  So
setting the listener on the capture phase instead works around this.
Probably would also work to dipsose the terminal.

code-asher committed:
Prevent issuing terminal reload when already reloading

I am not sure this actually causes any issues, but might as well.

code-asher committed:
Ignore modifier keys for reconnecting terminal

aslilac reviewed (commented):
I might be missing something, but it seems like this `status` isn't doing anything

code-asher commented on site/src/pages/TerminalPage/TerminalPage.tsx:
```diff
@@ -470,21 +470,62 @@ const useReloading = (isDisconnected: boolean) => {
+      if (isModifier(event)) {
+        modifierKeyState[event.key] = true;
+      } else if (!isModified()) {
+        setStatus("reloading");
```
Yeah I thought the same thing, it could depend on the browser or something, but in Chromium I tested via a `console.log` statement and it will happily run multiple times until the reload "kicks in".

code-asher replied:
That said, running `window.location.reload()` multiple times is probably harmless.

code-asher replied:
One more thing, setting the status here to `reloading` causes the overlay to display "Reloading...", so whatever `window.location.reload()` is doing, it appears to not quite be instant.

aslilac replied:
wow, ok 😅 I thought I'd even played with this before and touching `location` instantly killed your code

aslilac replied:
that said, it seems like this would benefit from being instant, rather than waiting for react to do an update cycle. maybe we should use `useRef` for this?

code-asher replied:
lol I know what you mean, I could have sworn I did the same!

code-asher replied:
Oh right, good point.  Maybe it would be nice if the UI could update immediately as well since if the reload happens too fast they might never see the `Reconnecting...` text?  Wonder if we can force an update or something.

code-asher replied:
I experimented a bit to see if I could get the keydown handler firing twice before React updates and the listener is removed but apparently I am not able to mash a key fast enough lol

I did find a bug though, if you ctrl+tab away then come back via click, the page still thinks ctrl is being pressed and nothing will reload until you press and release ctrl again.  Not sure if this can be solved with keydown...too bad it is not like the deprecated `keypress` event that would tell you if modifier keys were being held down.

code-asher replied:
Wait what no `keydown` does have that information.  What am I even doing??

code-asher commented on site/src/pages/TerminalPage/TerminalPage.tsx:
```diff
@@ -470,21 +470,62 @@ const useReloading = (isDisconnected: boolean) => {
 
-    document.addEventListener("keydown", keyDownHandler);
+    const isModifier = (event: KeyboardEvent): boolean => {
+      return typeof modifierKeyState[event.key] !== "undefined";
```
Oooo nice!

code-asher committed:
Swap undefined check for `in`

Co-authored-by: Kayla Washburn <mckayla@hey.com>

code-asher committed:
Use modifier properties on keydown event

Instead of manually tracking modifiers with keydown/keyup, as this can
miss keys like in the case of ctrl+tab, leaving ctrl in a permanent "on"
state.

The other modifiers do not appear to be used in conjunction with other
keys, except possibly Fn?  So this should be more or less the same
functionality.

aslilac reviewed (approved):
:+1:
